package/MDAnalysis/analysis/pca.py

```diff
@@ -253,18 +253,29 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
         """
         if not self._calculated:
             self.run()
-        # TODO has to accept universe or trajectory slice here
-        if self._atoms.atoms != atomgroup.atoms:
-            warnings.warn('This is a transform for different atom types.')
+
         if isinstance(atomgroup, Universe):
             atomgroup = atomgroup.atoms
 
+        if(self._n_atoms != atomgroup.n_atoms):
+            raise ValueError('PCA transform needs to be fit on same number of'
+                             ' atoms')
+        if not (self._atoms.types == atomgroup.types).all():
+            warnings.warn('This is a transform for different atom types.')
+
         traj = atomgroup.universe.trajectory
         start, stop, step = traj.check_slice_indices(start, stop, step)
-        dot = np.zeros((traj.n_frames, n_components))
+        n_frames = len(range(start, stop, step))
+
+        if n_components is None:
+            dim = self.p_components.shape[1]
+        else:
+            dim = n_components
```

Comment:
`dim = n_components if n_components is not None else self.p_components.shape[1]`